### PR TITLE
Add new command for discoverying function triggers.

### PR DESCRIPTION
**Pull request by taeold**
Adds an internal CLI command for discovering function triggers in a given project directory:

```
$ cd my-project
$ ls
firebase.json
functions/
$ firebase internaltesting:functions:discover
{
  "default": {
    "requiredAPIs": [],
    "endpoints": [
      {
        id: "helloWorld",
        platform: "gcfv2",
        ...
      },
      ...
    ]
  }
}
```

I intend to use the command to wire up integration test for discovery portion of a function deployment in [the follow up PR](https://github.com/firebase/firebase-tools/pull/5214).

This PR makes a small refactoring of the function deployment code to extract out the pieces for discovering functions. I verified that the function deploy integration test still passes after this refactor ([Test Run](https://github.com/firebase/firebase-tools/actions/runs/3397016650)).

DO NOT MERGE: New command requires an internal API proposal approval.

### CODE REVIEW

**Comment by taeold** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5213** <a href="https://app.graphite.dev/github/pr/firebase/firebase-tools/5213" target="_blank">Graphite</a>  👈
    * **PR #5214** <a href="https://app.graphite.dev/github/pr/firebase/firebase-tools/5214" target="_blank">Graphite</a> 
      * **PR #5215** <a href="https://app.graphite.dev/github/pr/firebase/firebase-tools/5215" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/firebase/firebase-tools/5213?utm_source=stack-comment).

**Comment by colerogers** [comment]
I would take the side of running this through API council since it has to be public facing and they might have naming suggestions for internal only commands. 

Alternatively, I think we should at least get @bkendall's sign-off as the CLI TL before merging this

**Comment by taeold** [comment]
@colerogers Since @bkendall won't be back for some time, I'll go ahead and draft an API proposal.

File: src/commands/functions-discover.ts
Comment (TheIronDev) [inline src/commands/functions-discover.ts:20]: Should this `JSON.stringify` be try/catch'd ?

Comment (taeold) [reply src/commands/functions-discover.ts:20]: I feel like that edge case is rare enough for us to just rely on the CLI's global try-catch mechanism. We are going to rethrow the error anyway right?

WDYT?

**Review by colerogers — APPROVED** [review]
🔥

File: src/commands/internaltesting-functions-discover.ts
Comment (bkendall) [inline src/commands/internaltesting-functions-discover.ts:12]:
```suggestion
  .description("discover function triggers defined in the current project directory")
```

File: src/deploy/functions/prepare.ts
Comment (bkendall) [inline src/deploy/functions/prepare.ts:422]: This is misleading. "for testing purposes only" very much means, in very other context: "I need to export this to write tests against it", *not* "this is exposed for an `internaltesting` command". Please clarify that it's for the command, rather than simply "testing".

File: src/commands/internaltesting-functions-discover.ts
Comment (bkendall) [inline src/commands/internaltesting-functions-discover.ts:16]: can this `as` be removed? (maybe `getProjectAdminSdkConfigOrCached` typed? if not, maybe add a TODO to clean up the `as`?

**Review by bkendall — CHANGES_REQUESTED** [review]

**Review by bkendall — APPROVED** [review]

**Comment by taeold** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/firebase/firebase-tools/5213?utm_source=stack-comment).


Job Reference: Uhty8UE38l3MKEIRA9gp

**Pull request merged**
